index.js

```diff
@@ -140,12 +140,20 @@ exports.on_local_interface = function (ip) {
 }
 
 exports.is_local_host = async function (host) {
+  const public_ip = await this.get_public_ip()
 
-  if (net.isIP(host)) return this.is_local_ip(host);
+  if (net.isIP(host)) {
+    if (public_ip && public_ip === host) return true
+    return this.is_local_ip(host)
+  }
 
   try {
     const ips = await this.get_ips_by_host(host)
-    return ips.length ? this.is_local_ip(ips[0]) : false
+    if (public_ip) ips.push(public_ip)
+    for (const ip of ips) {
+      if (this.is_local_ip(ip)) return true
+    }
```

Comment:
It seems to me that `is_local_ip` checks only loopback and local IPs. If that's correct, then this code can never return true when the given IP is the host public IP (which is the point of this PR). A  public IP is not a local IP. Or did I miss something?

Reply:
Actually, the first test in `is_local_ip` is `on_local_interface`. A local IP is a locally *bound* IP address, not necessarily a *loopback* IP.

Also, since the tests pass, does this mean your new tests don't actually exercise the feature you want?

Reply:
Thanks for the feedback @msimerson. Yes, this code doesn't exercise the feature I'm looking for.

A server public IP is not necessarily a locally bound IP. For instance in our case we run Haraka on AWS EC2 instances, and these have private IPs bound locally and mapped to external public IPs on AWS. I suppose the same would apply for a server behind a proxy.

For these cases, it's necessary to check if the target host IP is equal to the current host public IP. Something like [this](https://github.com/haraka/haraka-net-utils/pull/76/commits/2f46e0703ff99d3cc1bb4564931c7dbb2f9aca9c#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R147-R149) seemed to work in my tests.

Reply:
I think this is one of those situations where we need a comment, or some really well self-documenting code, that enumerates all the different ways `is_local_host` should pass. I'll start:

1. Destination hostname matches "my" Haraka hostname
2. Destination host resolves to an IP bound on a local interface
3. Destination host resolves to an IP that matches Haraka's public IP
4. ...

Reply:
4. Destination hostname equals to `get_primary_host_name`.
5. Destination hostname resolves to an IP that macthes `get_primary_host_name`'s IP.

I had not thought of that earlier. I'd say it would solve the issue also in the case of Haraka being behind a proxy or NLB.

Or another solution could be to fail the delivery when Haraka sees that the email has been looping to itself for X attempts? I suppose we could have a piece of code somewhere looking at the last few Received headers of the email, and if these are all something `is_local_host` or `get_primary_host_name` would return true for, fail the delivery to stop the loop.

What do you think? Shall we first add 4. and 5. matches to `is_local_host` and see how that goes?

Reply:
>  I suppose we could have a piece of code somewhere looking at the last few Received headers of the email, and if these are all something is_local_host or get_primary_host_name would return true for, fail the delivery to stop the loop.

There's already something like this in place that puts an end to mail loops. I forget exactly where.

This is feeling like there's enough checks that it needs a simplified solution. Would logic like this work?

a. assemble a list of local hostnames
b. assemble a list of local IPs
c. check if hostname is in the local hosts list
d. check if remote IP is in local IPs list

Reply:
Thanks @msimerson. I'd add these:

e. assemble a list of the IPs that the local hostnames resolve to (not necessarily the same thing as the local IPs if the server is behind a proxy or LB)
f. check if remote IP is in the local hostnames IPs

In case an hostname not in the local ones actually points to the same server. Not entirely sure if this is a setup that people use, but better to cover all cases.

Reply:
Any update on this @msimerson? I can have another go at it based on checks a. to f. if you'd like.

Reply:
Regarding your additions, I see `e.` as a subset of `b.`, for the same reasons we've covered. With that being the the case, the `f.` condition is already covered in `d.`.  Right?

Reply:
Thank you @msimerson. I guess it depends on wording. To me "local IPs" refers only to the IPs attached to the network interfaces.

In `e.` I add the IPs that local hostnames resolve to, which are not necessarily the same. For instance, on AWS an EC2 instance has a public IP but from whithin the instance only its private IP counterpart is associated to a network interface.

As long as we pick both the local IPs and IPs that the local hostnames resolve to for our `d.`/`f.` test, I think we're good.

Reply:
What about this @msimerson:

```
exports.is_local_host = async function (host) {
  let host_ips;
  let local_host_ips;

  try {
    host_ips = net.isIP(host) ? [host] : await this.get_ips_by_host(host);
    local_host_ips = await this.get_ips_by_host(os.hostname());
  } catch (e) {
    return false;
  }

  const non_local_ips = host_ips.filter(ip => !this.is_local_ip(ip) && !local_host_ips.includes(ip));
  return !non_local_ips.length;
}
```

That solves my use case and should cover most if not all the cases we talked about. 

I chose to have the function return `true` only if all of the remote IPs is are a local IPs, but that could be changed to returning `true` if at least one of the remote IPs is a local one, if you think that's preferable.

Reply:
Perhaps we should add the IP from `get_public_ip_async()` to `local_host_ips`. Some setups could have `os.hostname()` resolve to a different IP.

Reply:
Here's an updated version. I also replaced `os.hostname()` with `get_primary_host_name()`.

```
exports.is_local_host = async function (host) {
  let host_ips;
  let local_host_ips;

  try {
    host_ips = net.isIP(host) ? [host] : await this.get_ips_by_host(host);
    local_host_ips = [...await this.get_ips_by_host(this.get_primary_host_name()), await this.get_public_ip()];
  } catch (e) {
    return false;
  }

  const non_local_ips = host_ips.filter(ip => !this.is_local_ip(ip) && !local_host_ips.includes(ip));
  return !non_local_ips.length;
}
```